frontend/ui/widget/virtualkeyboard.lua

```diff
@@ -45,16 +47,60 @@ local VirtualKey = InputContainer:new{
 
 function VirtualKey:init()
     if self.keyboard.symbolmode_keys[self.label] ~= nil then
-        self.callback = function () self.keyboard:setLayout("Sym") end
+        self.callback = function () self.keyboard:setLayer("Sym") end
         self.skiptap = true
     elseif self.keyboard.shiftmode_keys[self.label] ~= nil then
-        self.callback = function () self.keyboard:setLayout("Shift") end
+        self.callback = function () self.keyboard:setLayer("Shift") end
         self.skiptap = true
     elseif self.keyboard.utf8mode_keys[self.label] ~= nil then
-        self.callback = function () self.keyboard:setLayout("IM") end
+        self.key_chars = self:genkeyboardLayoutKeyChars()
+        self.callback = function ()
+            local current = G_reader_settings:readSetting("keyboard_layout")
+            local keyboard_layouts = G_reader_settings:readSetting("keyboard_layouts")
```

Comment:
Like elsewhere in the code.
```suggestion
            local keyboard_layouts = G_reader_settings:readSetting("keyboard_layouts") or {}
```

Reply:
fixed